File: backend/src/appointment/appointment.policy.ts
export function calculateEnd(start: Date, serviceDurations: number[]): Date {
  const totalMinutes = serviceDurations.reduce((acc, d) => acc + d, 0);
  return new Date(start.getTime() + totalMinutes * 60 * 1000);
}

export function canCancel(appointmentStart: Date, now: Date): boolean {
  const twoHoursBefore = new Date(appointmentStart.getTime() - 2 * 60 * 60 * 1000);
  return twoHoursBefore >= now;
}

export function cashbackEarned(amountPaid: import('@prisma/client').Prisma.Decimal, rate: number): import('@prisma/client').Prisma.Decimal {
  const { Prisma } = require('@prisma/client');
  return new Prisma.Decimal(amountPaid).mul(rate).toDecimalPlaces(2, Prisma.Decimal.ROUND_HALF_UP);
}

File: backend/src/main/java/com/razorfy/appointment/AppointmentPolicy.java
package com.razorfy.appointment;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.OffsetDateTime;
import java.util.List;

public final class AppointmentPolicy {

    private AppointmentPolicy() {}

    public static OffsetDateTime calculateEnd(OffsetDateTime start, List<Integer> serviceDurations) {
        int totalMinutes = serviceDurations.stream().mapToInt(Integer::intValue).sum();
        return start.plusMinutes(totalMinutes);
    }

    public static boolean canCancel(OffsetDateTime appointmentStart, OffsetDateTime now) {
        return !appointmentStart.minusHours(2).isBefore(now);
    }

    public static BigDecimal cashbackEarned(BigDecimal amountPaid, BigDecimal rate) {
        return amountPaid.multiply(rate).setScale(2, RoundingMode.HALF_UP);
    }
}
